control/autoware_control_toolbox/include/utils_act/act_utils.hpp

```diff
@@ -0,0 +1,621 @@
+  typename T, typename std::enable_if_t<std::is_floating_point_v<T>, std::vector<T>> * = nullptr>
+
+std::vector<T> operator*(T const &a, std::vector<T> const &vec)
+{
+  std::vector<T> temp{vec};
+
+  std::transform(temp.begin(), temp.end(), temp.begin(), [&](auto &xi)
+  { return xi * a; });
+
+  //	for (auto& x: temp)
+  //	{
+  //		x = x * a;
+  //	}
+
+  return temp;
+}
+
+template<
+  typename T, typename std::enable_if_t<std::is_floating_point_v<T>, std::vector<T>> * = nullptr>
+
+std::vector<T> &operator*=(T const &a, std::vector<T> &vec)
+{
+  std::transform(vec.begin(), vec.end(), vec.begin(), [&](auto &xi)
+  { return xi * a; });
+
+  //	for (auto& x: vec)
+  //	{
+  //		x = x * a;
+  //	}
+
+  return vec;
+}
+
+namespace ns_utils
+{
+/**
+ * @brief signum function.
+ * */
+template<typename T>
+int sgn(T val)
+{
+  return (T(0) < val) - (val < T(0));
+}
+
+// Comparing data types.
+template<class T, typename std::enable_if_t<std::is_integral_v<T>, bool> * = nullptr>
+bool isEqual(T a, T b)
+{
+  return a == b;
+}
+
+template<class T, typename std::enable_if_t<std::is_floating_point_v<T>, bool> * = nullptr>
+bool isEqual(T a, T b)
+{
+  return abs(a - b) < std::numeric_limits<T>::epsilon();
+}
+
+/**
+ * Linear extrapolation
+ * */
+template<typename T, typename std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+void extrapolate(
+  std::vector<T> const &tbase, std::vector<T> const &ybase, T const &tnew, T &ynew)
+{
+  if (tnew < tbase[0])
+  {
+    auto const &t0 = tbase[0];
+    auto const &t1 = tbase[1];
+
+    auto const &y0 = ybase[0];
+    auto const &y1 = ybase[1];
+
+    auto const &ratio = (t0 - tnew) / (t1 - t0);
+    ynew = y0 - ratio * (y1 - y0);
+  }
+
+  if (tnew > tbase.back())
+  {
+    auto const &tn = tbase.rbegin()[0];
+    auto const &tn_1 = tbase.rbegin()[1];
+
+    auto const &yn = ybase.rbegin()[0];
+    auto const &yn_1 = ybase.rbegin()[1];
+
+    auto const &ratio = (tnew - tn) / (tn - tn_1);
+    ynew = yn + ratio * (yn - yn_1);
+  }
+}
+
+/**
+ * @brief Saturates given values
+ * */
+template<typename T>
+constexpr T clamp(const T &val, const T &lower, const T &upper)
+{
+  return std::max(lower, std::min(val, upper));
+}
+
+template<typename T, typename std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+T deg2rad(T const &angle)
+{
+  return angle * M_PI / 180.;
+}
+
+/**
+ * brief abs_diff for size_t
+ * */
+template<typename T>
+T abs_diff(T a, T b)
+{
+  return a > b ? a - b : b - a;
+}
+
+/**
+ * @brief angle wrapping w(x) = mod(x, +pi, 2*pi) - pi
+ * */
+template<typename T, typename std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+constexpr T angleDistance(T const &theta)
+{
+  T mod_theta_2pi = std::fmod(theta + M_PI, 2 * M_PI) - M_PI;
+
+  return mod_theta_2pi < -M_PI ? mod_theta_2pi + 2 * M_PI : mod_theta_2pi;
+}
+
+template<typename T, typename std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+constexpr T angleDistance(T const &theta, T const &theta_ref)
```

Comment:
This can probably be replaced by https://github.com/autowarefoundation/autoware.universe/blob/main/common/motion_common/src/motion_common/trajectory_common.cpp#L36-L39